lib/Backends/NNPI/InferencePool.cpp

```diff
@@ -79,22 +83,44 @@ void InferenceThreadEnv::execute(RunIdentifierTy runId,
                                  resultCB);
     auto *t = ioTensors_.at(in.first);
 
-    switch (t->getElementType()) {
-    case glow::ElemKind::Int64ITy: {
-      // Convert int64_t tensors to int32.
-      int64_t *pInput = reinterpret_cast<int64_t *>(t->getUnsafePtr());
-      const size_t unpaddedSize = t->getUnpaddedSizeInBytes() / sizeof(int64_t);
-      int32_t *tmp = new int32_t[unpaddedSize];
-      for (size_t i = 0; i < unpaddedSize; i++) {
+    const bool downcastInt64 = t->getElementType() == glow::ElemKind::Int64ITy;
+    size_t paddedSize = t->getSizeInBytes();
+    size_t unpaddedSize = t->getUnpaddedSizeInBytes();
+    if (downcastInt64) {
+      paddedSize /= 2;
+      unpaddedSize /= 2;
+    }
+    char *bufferPtr = t->getUnsafePtr();
+
+    // Convert int64 tensors to int32.
+    if (downcastInt64) {
+      int64_t *pInput = reinterpret_cast<int64_t *>(bufferPtr);
+      const size_t hostElems = unpaddedSize / sizeof(int32_t);
+      int32_t *tmp = new int32_t[hostElems];
+      for (size_t i = 0; i < hostElems; i++) {
         tmp[i] = static_cast<int32_t>(pInput[i]);
       }
-      rawInputs_.push_back(tmp);
+      bufferPtr = reinterpret_cast<char *>(tmp);
+      tmpBuffers_.insert(bufferPtr);
+    }
+
+    // If this tensor cannot be treated as partial but it is partial, then we
+    // need to make a copy of the underlying buffer to the correct unpadded
+    // size, and zero out anything that's leftover in the extra padding, as it
+    // is still going to be used.
+    if (!partialTensorInputs_.count(t) && paddedSize != unpaddedSize) {
+      char *tmp = new char[paddedSize];
+      // Copy over original data.
+      memcpy(tmp, bufferPtr, unpaddedSize);
+      // Zero out rest of the new buffer.
+      memset(tmp + unpaddedSize, 0, paddedSize - unpaddedSize);
```

Comment:
Zero isn't always the idempotent value (e.g. 1 for ReduceMul).
You can extend the mechanism in the util function to also provide the correct value to pad with.
Not sure if there's value in that for your usage and either way can be added later.

Reply:
Alternatively, if the desired behavior is to always use 0 for padding then the memset can happen above the backend (shared by all backends) and that tensor can get its unpaddedSize set to the full size at that point - thus removing the need for any special handling in the backends.